### PR TITLE
Updated for API 7 support

### PR DESCRIPTION
**Pull request by codeHusky**
Untested but should work

### CODE REVIEW

**Comment by ryantheleach** [comment]
@codeHusky Close, but this will call the CauseStackManager async,

https://gist.github.com/ryantheleach/1d03f5e2aa643591bed87462eb8b3221 

Constructing the cause and firing it yourself is required for async tasks.

**Comment by xKnightGamingx** [comment]
https://gist.github.com/xKnightGamingx/230dbae027d18081afa6c425d2abc8a1

**Comment by codeHusky** [comment]
Fixed the issue @ryantheleach @xKnightGamingx

File: sponge/pom.xml [hunk 1]
```diff
@@ -42,20 +42,28 @@
...
+            </plugin>
         </plugins>
     </build>
 
     <dependencies>
         <dependency>
             <groupId>com.vexsoftware</groupId>
             <artifactId>nuvotifier-common</artifactId>
-            <version>${project.parent.version}</version>
+            <version>2.3.5</version>
```
Comment (Ichbinjoe) [inline sponge/pom.xml]: Why was this changed?

**Comment by Ichbinjoe** [comment]
@codeHusky Thanks for making these changes. Please address my comment, and I will get this merged in.

**Comment by ryantheleach** [comment]
@Ichbinjoe Huskies changes change it from being processed explicitly async with the vote as the cause, to 'maybe' async, without a vote as a cause. (maybe, because I'm not sure of how the votes are received)

It's up to you if you merge as is, but I suspect you were running it async to begin with for good reason? (If not, I highly recommend at least processing the events synchronous in order to get bad behaving plugins in line)

Regardless of the method of the event being created async, the cause tracker can not be called asynchronously, so you will either need to make a new cause (like my patch in my comment) Or get the cause in a sync task, then throw the event async/sync as you please. (It's my understanding causes are immutable)

**Comment by Ichbinjoe** [comment]
Lets make it forced in sync then. Tux originally wrote the code for sponge, so I have no real insight into why the event is being explicitly called async. If I were to do this over, I would have called this event in sync - this is consistent with how bukkit handles the vote events (iirc), and relieves much of the threaded consideration from plugin authors. However, we have set a precedent that this event will be called async, and changing it to sync may have serious considerations for plugin authors who have already assumed it is fired off async. If others have opinions about how this should be done, please comment. Thanks for catching this @ryantheleach, I have never touched sponge and appreciate when people catch things like this.

File: sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java [hunk 2]
```diff
@@ -366,15 +359,9 @@ public void onForward(final Vote v) {
...
         }
-        Sponge.getScheduler().createTaskBuilder()
-                .execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        VotifierEvent event = new VotifierEvent(v, Sponge.getCauseStackManager().getCurrentCause());
-                        Sponge.getEventManager().post(event);
-                    }
-                })
-                .async()
```
Comment (Ichbinjoe) [inline sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: View discussion in PR thread - we need to decide if this is async or sync, and keep it explicit. At the moment, this will call it async within a netty io thread iirc.

Comment (ryantheleach) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: The point is more the call to the CauseStackManager which will actively throw exceptions if called asynchronously.

Comment (Ichbinjoe) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: I understand that ryan, however, the code change also removes the explicit wrapping. I would like that added back in so that we are scheduling it within a sponge scheduler loop, and on top of that doing it explicitly instead of just running it in netty. I agree that the CauseStackManager should be called in sync since it doesn't like not being called in that way, however, we need to be explicit on how we send the event out to the subscribers of the event.

File: sponge/pom.xml (hunk 1 above)
Comment (codeHusky) [reply sponge/pom.xml]: Accidental, was trying to figure out how the project worked.

File: sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java (hunk 2 above)
Comment (codeHusky) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: There isn't too much of an actual reason to call it asynchronously. I can't see a plugin that couldn't handle the event being fired *synchronously*. I'd expect issues the other way around, to be honest.

Comment (Ichbinjoe) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: I would expect issues called async as well - I am very unsure why Tux made the original decision to fire this off async. If we aren't concerned about a possible dependencies on the listeners, then lets wrap this and call it in sync. I'll be sure to add a release note noting this - I don't see issues with people doing weird stuff depending on an async event call, but you never know.

Comment (codeHusky) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: The current code is apparently having some alleged issues and failing to receive votes but I’ll test it ASAP and see if that’s true

Comment (LoXR) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: Will this fix the "[Votifier] Unable to process vote from" problem in console. Tried to use Nu in Bungee to have it work across multiple servers.

Comment (Keaton1188) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: Hopefully!

Comment (Ichbinjoe) [reply sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java]: @codeHusky @LoXR I may have fixed the voting issue, netty had an issue where it acted squirrelish with a class we were using from decoding. I would love to PR this as soon as I can so I can get 2.3.6 out the door (probably should be 2.4.0, but whatevs)

**Comment by Scarsz** [comment]
Looks like this has gone stale so I did some of my own testing for anyone wanting a fix like I was. Currently, @codeHusky's fork works with API 7 but with an issue of enforced thread safety.

Replacing lines [343 & 344](https://github.com/codeHusky/NuVotifier/blob/master/sponge/src/main/java/com/vexsoftware/votifier/sponge/VotifierPlugin.java#L343-L344) with the following ensures that the event is posted on the main thread and everything works well.

```java
Task.builder().execute(task -> {
    VotifierEvent event = new VotifierEvent(vote, Sponge.getCauseStackManager().getCurrentCause());
    Sponge.getEventManager().post(event);
}).submit(this);
```

Results of a [test](https://www.craftservers.net/tools/votifier-tester/) and listener to kill my IGN:

```java
[01:45:14] [nioEventLoopGroup-2-1/INFO] [nuvotifier]: Got a protocol v1 vote record from /x.x.x.x:35902 -> Vote (from:CraftServers.net username:Notch address:x.x.x.x timeStamp:2018-01-07 06:45:11.602732)
[01:45:14] [Server thread/INFO] [nucleus]: Server ran the command: /kill Scarsz
[01:45:14] [Server thread/INFO]: Scarsz died
[01:45:15] [Server thread/INFO]: You killed Scarsz.
```

**Comment by KhaozFrost** [comment]
still causing issues in latest sponge api7. always spams unable to process vote but thats the only logs given.
`10.01 09:54:17 [Server] nioEventLoopGroup-3-1/ERROR [nuvotifier]: Unable to process vote from /127.0.0.1:57452 10.01 09:54:18 [Server] nioEventLoopGroup-3-1/ERROR [nuvotifier]: Unable to process vote from /127.0.0.1:57472 10.01 09:54:20 [Server] nioEventLoopGroup-3-1/ERROR [nuvotifier]: Unable to process vote from /127.0.0.1:57478 10.01 09:54:23 [Server] nioEventLoopGroup-3-1/ERROR [nuvotifier]: Unable to process vote from /127.0.0.1:57498 10.01 09:54:28 [Connect] User FrostEffects, IP unknown 10.01 09:54:28 [Multicraft] FrostEffects ran command Message of the Day 10.01 09:54:28 [Server] Server thread/INFO Multicraft compatibility has been enabled. To use the Nucleus list command from the console, use /ls. 10.01 09:54:28 [Server] nioEventLoopGroup-3-1/ERROR [nuvotifier]: Unable to process vote from /127.0.0.1:57532 10.01 09:54:36 [Server] nioEventLoopGroup-3-1/ERROR [nuvotifier]: Unable to process vote from /127.0.0.1:57590`

**Comment by Ichbinjoe** [comment]
There has been no activity from the original author in some time. I have manually merged this in with the suggested changes.
